calculator_test.go

```diff
@@ -2,23 +2,200 @@ package calculator_test
 
 import (
 	"calculator"
+	"fmt"
+	"math/rand"
 	"testing"
 )
 
+type testCase struct {
+	a, b        float64
+	want        float64
+	name        string
+	errExpected bool
+}
+
+type variadicTestCase struct {
+	nums        []float64
+	want        float64
+	name        string
+	errExpected bool
+}
+
+type evaluateTestCase struct {
+	expression  string
+	want        float64
+	errExpected bool
+}
+
 func TestAdd(t *testing.T) {
 	t.Parallel()
-	var want float64 = 4
-	got := calculator.Add(2, 2)
-	if want != got {
-		t.Errorf("want %f, got %f", want, got)
+	testCases := []variadicTestCase{
+		{nums: []float64{1, 2, 3}, want: 6, name: "1+2+3 = 6"},
+		{nums: []float64{1, 2, 3, 4}, want: 10, name: "1+2+3+4 = 10"},
+	}
+
+	for _, tc := range testCases {
+		got := calculator.Add(tc.nums...)
+
+		if tc.want != got {
+			t.Errorf("Error: unexpected value returned - want %.1f, got %.1f", tc.want, got)
+		}
+	}
+}
+
+func TestAddRandom(t *testing.T) {
+	t.Parallel()
+	for i := 0; i < 10000000; i++ {
+		a := rand.Float64()
+		b := rand.Float64()
+		want := a + b
+		got := calculator.Add(a, b)
+
+		if want != got {
+			t.Errorf("Want: %.2f, got %.2f", want, got)
+		}
 	}
 }
 
-// func TestSubtract(t *testing.T) {
-// 	t.Parallel()
-// 	var want float64 = 2
-// 	got := calculator.Subtract(4, 2)
-// 	if want != got {
-// 		t.Errorf("want %f, got %f", want, got)
-// 	}
-// }
+func TestSubtract(t *testing.T) {
+	t.Parallel()
+
+	testCases := []variadicTestCase{
+		{nums: []float64{3, 2, 1}, want: 0, name: "3 - 2 - 1 = 0"},
+		{nums: []float64{100, 8, 12}, want: 80, name: "100 - 8 - 12 = 10"},
+		{nums: []float64{1, 8, 10}, want: -17, name: "1 - 8 - 10 = -17"},
+	}
+
+	for _, tc := range testCases {
+		got := calculator.Subtract(tc.nums...)
+
+		if tc.want != got {
+			t.Errorf("Error: unexpected value returned - want %.1f, got %.1f", tc.want, got)
+		}
+	}
+}
+
+func TestMultiply(t *testing.T) {
+	t.Parallel()
+
+	testCases := []variadicTestCase{
+		{nums: []float64{3, 2, 1}, want: 6, name: "3 * 2 * 1 = 6"},
+		{nums: []float64{100, 8, 12}, want: 9600, name: "100 * 8 * 12 = 9600"},
+		{nums: []float64{-2, 8, 10}, want: -160, name: "-1 * 8 * 10 = -160"},
+		{nums: []float64{400, 13, 0, 4}, want: 0, name: "400 * 13 * 0 * 4 = 0"},
+	}
+
+	for _, tc := range testCases {
+		got := calculator.Multiply(tc.nums...)
+
+		if tc.want != got {
+			t.Errorf("Error: unexpected value returned - want %.1f, got %.1f", tc.want, got)
+		}
+	}
+
+}
+
+func TestDivide(t *testing.T) {
+	t.Parallel()
+
+	testCases := []variadicTestCase{
+		{nums: []float64{7, 0, 10, 44}, want: 999, name: "1 / 8 / -10 = -0.0125", errExpected: true},
+		{nums: []float64{3, 2}, want: 1.5, name: "3 / 2 = 1.5", errExpected: false},
```

Comment:
Also try `2, 3` for an interesting test case!

Reply:
Is that interesting because of how the rounding works in float64s, or is there another reason? 🤔

(added it to a902892)

Reply:
It's interesting because `want != got` is likely to produce the wrong result.

Reply:
👍